File: go/toss_stream_test.go
package hermetix

// 토스증권 스트림 — Bearer 핸드셰이크, 선언형 구독(합치기·거부 제외), 체결·호가·주문 이벤트(누적 델타), PING 하트비트, 재접속, 픽스처(AsyncAPI 샘플) 파싱.

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/coder/websocket"
)

type tossServerConn struct {
	*serverConn
	authorization string
}

// newTossWsServer - 업그레이드 요청의 Authorization 헤더를 함께 기록한다.
func newTossWsServer(t *testing.T) (*httptest.Server, chan *tossServerConn) {
	t.Helper()
	connections := make(chan *tossServerConn, 8)
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		auth := r.Header.Get("Authorization")
		conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{InsecureSkipVerify: true})
		if err != nil {
			return
		}
		sc := &tossServerConn{serverConn: &serverConn{conn: conn, received: make(chan string, 64)}, authorization: auth}
		connections <- sc
		for {
			kind, data, err := conn.Read(context.Background())
			if err != nil {
				return
			}
			if kind == websocket.MessageText {
				sc.received <- string(data)
			}
		}
	}))
	t.Cleanup(srv.Close)
	return srv, connections
}

func waitTossConn(t *testing.T, connections chan *tossServerConn) *tossServerConn {
	t.Helper()
	select {
	case c := <-connections:
		return c
	case <-time.After(10 * time.Second):
		t.Fatal("접속이 오지 않음")
		return nil
	}
}

func newTossTestStream(t *testing.T) (*TossMarketStream, chan *tossServerConn) {
	t.Helper()
	srv, connections := newTossWsServer(t)
	stream := newTossMarketStream(wsURL(srv), func() (string, error) { return "ACCESS", nil }, func() (string, error) { return "3", nil })
	stream.declareDelay = 50 * time.Millisecond
	t.Cleanup(func() { stream.Close() })
	return stream, connections
}

type tossDeclaration []map[string]any

func parseDeclaration(t *testing.T, raw string) tossDeclaration {
	t.Helper()
	var d tossDeclaration
	if err := json.Unmarshal([]byte(raw), &d); err != nil {
		t.Fatalf("declaration is not a JSON array: %s", raw)
	}
	return d
}

func (d tossDeclaration) codes(typ string) []string {
	for _, item := range d {
		if item["type"] == typ {
			raw, _ := item["codes"].([]any)
			out := make([]string, 0, len(raw))
			for _, c := range raw {
				out = append(out, c.(string))
			}
			return out
		}
	}
	return nil
}

func TestTossStreamBearerAndDeclaration(t *testing.T) {
	stream, connections := newTossTestStream(t)
	stream.SubscribeTrades([]string{"KRX:005930", "US:AAPL"}, func(TradeTick) {})
	_ = stream.SubscribeOrderBook([]string{"KRX:005930"}, func(OrderBookTick) {})
	_ = stream.SubscribeOrderEvents(func(OrderEvent) {})
	stream.Connect()

	conn := waitTossConn(t, connections)
	if conn.authorization != "Bearer ACCESS" {
		t.Fatalf("Authorization = %q", conn.authorization)
	}
	decl := parseDeclaration(t, conn.take(t))
	if decl[0]["id"] == nil || len(decl) != 5 {
		t.Fatalf("declaration = %v", decl)
	}
	if got := decl.codes("trade:kr"); len(got) != 1 || got[0] != "005930" {
		t.Fatalf("trade:kr = %v", got)
	}
	if got := decl.codes("trade:us"); len(got) != 1 || got[0] != "AAPL" {
		t.Fatalf("trade:us = %v", got)
	}
	if got := decl.codes("orderbook:kr"); len(got) != 1 || got[0] != "005930" {
		t.Fatalf("orderbook:kr = %v", got)
	}
	if got := decl.codes("personal:order"); len(got) != 1 || got[0] != "3" {
		t.Fatalf("personal:order = %v", got)
	}

	// 거부된 target 은 다음 선언에서 빠진다 (늦은 구독이 재선언을 유발)
	conn.send(t, `{"type":"subscriptions","id":"req-1","subscribed":["trade:kr:005930"],"rejected":[{"target":"trade:us:AAPL","code":"stock-not-found","message":"x"}]}`)
	time.Sleep(100 * time.Millisecond)
	stream.SubscribeTrades([]string{"000660"}, func(TradeTick) {})
	next := parseDeclaration(t, conn.take(t))
	if got := next.codes("trade:us"); len(got) != 0 {
		t.Fatalf("거부된 AAPL 이 다시 선언됐다: %v", got)
	}
	if got := next.codes("trade:kr"); len(got) != 2 || got[0] != "000660" || got[1] != "005930" {
		t.Fatalf("trade:kr = %v", got)
	}
}

func TestTossStreamDataFrames(t *testing.T) {
	fx := loadStreamFixture(t, "toss")
	stream, connections := newTossTestStream(t)
	ticks := make(chan TradeTick, 8)
	books := make(chan OrderBookTick, 8)
	events := make(chan OrderEvent, 8)
	stream.SubscribeTrades([]string{"US:AAPL", "KRX:005930", "000660"}, func(tick TradeTick) { ticks <- tick })
	_ = stream.SubscribeOrderBook([]string{"KRX:005930"}, func(b OrderBookTick) { books <- b })
	_ = stream.SubscribeOrderEvents(func(e OrderEvent) { events <- e })
	stream.Connect()
	conn := waitTossConn(t, connections)
	conn.take(t) // 선언

	conn.send(t, `{"type":"pong"}`)
	conn.send(t, `{"type":"error","error":{"code":"rate-limit-exceeded","message":"slow down"}}`)
	conn.send(t, fx.Frames[0])
	tick := waitTick(t, ticks)
	if tick.Symbol != "US:AAPL" || tick.Price.String() != "243.26" || tick.Quantity.String() != "8" || tick.CumulativeVolume != nil || tick.Change != nil {
		t.Fatalf("tick = %+v", tick)
	}
	if got := tick.Timestamp.In(kst).Format("15:04"); got != "23:30" {
		t.Fatalf("time = %s", got)
	}
	// 접두 없이 구독한 심볼은 접두 없이 돌려준다 (요청 표기는 채널 공통 — 같은 종목을 다른 표기로 두 번 구독하면 첫 표기를 쓴다)
	conn.send(t, `{"type":"message","topic":"trade:kr:000660","data":{"price":"71500","volume":"3","timestamp":"2026-09-14T10:00:00.000+09:00","currency":"KRW"}}`)
	if bare := waitTick(t, ticks); bare.Symbol != "000660" {
		t.Fatalf("bare tick = %+v", bare)
	}

	conn.send(t, fx.OrderBook.Frames[0])
	var book OrderBookTick
	select {
	case book = <-books:
	case <-time.After(5 * time.Second):
		t.Fatal("호가가 오지 않음")
	}
	best, _ := book.BestAsk()
	bid, _ := book.BestBid()
	if book.Symbol != "KRX:005930" || best.Price.String() != "71500" || best.Quantity.String() != "5" || bid.Price.String() != "71400" || book.TotalAskQuantity != nil {
		t.Fatalf("book = %+v", book)
	}

	take := func() OrderEvent {
		select {
		case e := <-events:
			return e
		case <-time.After(5 * time.Second):
			t.Fatal("이벤트가 오지 않음")
			return OrderEvent{}
		}
	}
	order := `{"orderId":"ORD-1","symbol":"AAPL","side":"BUY","orderType":"LIMIT","timeInForce":"DAY","status":"%s","price":"100.5","quantity":"10","currency":"USD","orderedAt":"2026-06-23T09:30:00.000+09:00","canceledAt":null,"execution":{"filledQuantity":"%s","averageFilledPrice":"100","filledAmount":null,"commission":null,"tax":null,"settlementDate":null}}`
	frame := func(event, status, filled string) string {
		return `{"type":"message","topic":"personal:order:3","data":{"event":"` + event + `","accountSeq":"3","order":` + fmt.Sprintf(order, status, filled) + `}}`
	}
	conn.send(t, frame("PENDING", "PENDING", "0"))
	if accepted := take(); accepted.Type != OrderAccepted || accepted.Symbol != "US:AAPL" || *accepted.Side != Buy || accepted.Quantity.String() != "10" {
		t.Fatalf("accepted = %+v", accepted)
	}
	conn.send(t, frame("CANCELING", "PENDING_CANCEL", "0")) // 중간 상태 — 무시
	conn.send(t, frame("PARTIAL_FILL", "PARTIAL_FILLED", "4"))
	partial := take()
	if partial.Type != OrderFilled || partial.Quantity.String() != "4" || partial.Price.String() != "100" || partial.RemainingQuantity.String() != "6" {
		t.Fatalf("partial = %+v", partial)
	}
	conn.send(t, frame("FILL", "FILLED", "10"))
	full := take()
	if full.Type != OrderFilled || full.Quantity.String() != "6" || full.RemainingQuantity.String() != "0" {
		t.Fatalf("full = %+v", full)
	}
	conn.send(t, frame("CANCEL_REJECTED", "CANCEL_REJECTED", "10"))
	if rejected := take(); rejected.Type != OrderRejected || rejected.Reason != "CANCEL_REJECTED" {
		t.Fatalf("rejected = %+v", rejected)
	}
}

func TestTossStreamHeartbeatAndReconnect(t *testing.T) {
	srv, connections := newTossWsServer(t)
	stream := newTossMarketStream(wsURL(srv), func() (string, error) { return "ACCESS", nil }, func() (string, error) { return "3", nil })
	stream.declareDelay = 50 * time.Millisecond
	stream.reconnectingWebSocket.heartbeat = 150 * time.Millisecond
	defer stream.Close()
	stream.SubscribeTrades([]string{"KRX:005930"}, func(TradeTick) {})
	stream.Connect()

	conn := waitTossConn(t, connections)
	parseDeclaration(t, conn.take(t))
	if ping := conn.take(t); ping != "PING" {
		t.Fatalf("heartbeat = %q", ping)
	}

	_ = conn.conn.Close(1000, "server-shutdown")
	second := waitTossConn(t, connections)
	if second.authorization != "Bearer ACCESS" {
		t.Fatalf("재접속 Authorization = %q", second.authorization)
	}
	if again := parseDeclaration(t, second.take(t)); len(again.codes("trade:kr")) != 1 {
		t.Fatalf("재선언 = %v", again)
	}
}

func TestTossTopicKey(t *testing.T) {
	for symbol, want := range map[string]string{"005930": "kr:005930", "KRX:005930": "kr:005930", "US:aapl": "us:AAPL"} {
		if got, err := TossTopicKey(symbol); err != nil || got != want {
			t.Fatalf("%s → %s (%v), expected %s", symbol, got, err, want)
		}
	}
	if _, err := TossTopicKey("JP:7203"); err == nil {
		t.Fatal("미지원 시장은 오류여야 한다")
	}
}

func TestTossStreamFixture(t *testing.T) {
	fx := loadStreamFixture(t, "toss")
	if fx.Measured {
		t.Fatal("toss 픽스처는 문서 기반(measured=false)이어야 한다")
	}
	// 체결: 누적거래량·등락이 없어(null) 공통 assertStreamTicks 대신 필드별 확인
	var frame struct {
		Topic string          `json:"topic"`
		Data  json.RawMessage `json:"data"`
	}
	if err := json.Unmarshal([]byte(fx.Frames[0]), &frame); err != nil {
		t.Fatal(err)
	}
	tick, ok := ParseTossTrade(frame.Topic, frame.Data)
	e := fx.Expected[0]
	if !ok || tick.Symbol != e.Symbol || tick.Price.String() != e.Price || tick.Quantity.String() != e.Quantity || tick.CumulativeVolume != nil || tick.Change != nil || tick.ChangeRate != nil {
		t.Fatalf("tick = %+v, expected %+v", tick, e)
	}
	if got := tick.Timestamp.In(kst).Format("15:04"); got != e.Time {
		t.Fatalf("time = %s, expected %s", got, e.Time)
	}

	if err := json.Unmarshal([]byte(fx.OrderBook.Frames[0]), &frame); err != nil {
		t.Fatal(err)
	}
	book, ok := ParseTossOrderBook(frame.Topic, frame.Data)
	be := fx.OrderBook.Expected[0]
	if !ok || book.Symbol != be.Symbol || len(book.Asks) != len(be.Asks) || len(book.Bids) != len(be.Bids) ||
		book.Asks[0].Price.String() != be.Asks[0].Price || book.Asks[0].Quantity.String() != be.Asks[0].Quantity ||
		book.Bids[0].Price.String() != be.Bids[0].Price || book.TotalAskQuantity != nil {
		t.Fatalf("book = %+v, expected %+v", book, be)
	}
	if got := book.Timestamp.In(kst).Format("15:04"); got != be.Time {
		t.Fatalf("book time = %s, expected %s", got, be.Time)
	}

	if err := json.Unmarshal([]byte(fx.OrderEvents.Frames[0]), &frame); err != nil {
		t.Fatal(err)
	}
	event, ok := ParseTossOrderEvent(frame.Data, nil)
	ee := fx.OrderEvents.Expected[0]
	if !ok || event.OrderID != ee.OrderID || string(event.Type) != ee.Type || event.Symbol != ee.Symbol || event.Side == nil || string(*event.Side) != ee.Side {
		t.Fatalf("event = %+v, expected %+v", event, ee)
	}
	assertDecimalEqual(t, "quantity", event.Quantity, ee.Quantity)
	assertDecimalEqual(t, "price", event.Price, ee.Price)
	assertDecimalEqual(t, "remaining", event.RemainingQuantity, ee.RemainingQuantity)
	if got := event.Timestamp.In(kst).Format("15:04"); got != ee.Time {
		t.Fatalf("event time = %s, expected %s", got, ee.Time)
	}
}

File: js/tests/tossStream.test.ts
/** 토스증권 실시간 스트림 — 로컬 ws 서버로 AsyncAPI 프로토콜을 재생한다 (Kotlin TossMarketStreamTest 대응, 문서 기반). */
import assert from "node:assert/strict";
import { test } from "node:test";
import { Decimal } from "decimal.js";
import { TossMarketStream, parseTossOrderBook, parseTossOrderEvent, parseTossTrade, tossTopicKey } from "../src/brokers/tossStream.js";
import type { OrderBookTick, OrderEvent, TradeTick } from "../src/models.js";
import { Queue, assertBooks, assertEvents, assertTrades, sleep, streamFixture, withServer } from "./wsHarness.js";

const fx = streamFixture("toss");
const TRADE_US = fx.frames[0];
const ORDERBOOK_KR = fx.orderBook.frames[0];
const ORDER_FILL = fx.orderEvents.frames[0];
const TRADE_KR = '{"type":"message","topic":"trade:kr:005930","data":{"price":"71500","volume":"3","timestamp":"2026-06-18T23:30:00.000+09:00","currency":"KRW"}}';
const opts = (url: string, extra: Partial<{ heartbeatMs: number; declareDelayMs: number }> = {}) =>
  ({ wsUrl: url, token: async () => "TOKEN", accountSeq: async () => "3", declareDelayMs: 50, ...extra });

const orderFrame = (event: string, filled: string, price = "100.5", avg = "100") =>
  `{"type":"message","topic":"personal:order:3","data":{"event":"${event}","accountSeq":"3","order":{"orderId":"ORD-1","symbol":"AAPL","side":"BUY","orderType":"LIMIT","timeInForce":"DAY","status":"x","price":"${price}","quantity":"10","currency":"USD","orderedAt":"2026-06-23T09:30:00.000+09:00","canceledAt":null,"execution":{"filledQuantity":"${filled}","averageFilledPrice":"${avg}"}}}}`;

test("toss: 핸드셰이크에 Bearer 토큰을 싣고, 구독 전체를 배열 하나로 선언한다", () =>
  withServer(async (url, connections) => {
    const stream = new TossMarketStream(opts(url));
    try {
      stream.subscribeTrades(["KRX:005930"], () => {});
      stream.subscribeTrades(["US:AAPL"], () => {});
      stream.subscribeOrderBook(["KRX:005930"], () => {});
      stream.subscribeOrderEvents(() => {});
      stream.connect();
      const conn = await connections.take();
      assert.equal(conn.header("authorization"), "Bearer TOKEN");
      const declaration = JSON.parse(await conn.received.take());
      assert.ok(Array.isArray(declaration));
      assert.equal(declaration[0].id, "req-1");
      const byType = Object.fromEntries(declaration.slice(1).map((d: { type: string; codes: string[] }) => [d.type, d.codes]));
      assert.deepEqual(byType["trade:kr"], ["005930"]);
      assert.deepEqual(byType["trade:us"], ["AAPL"]);
      assert.deepEqual(byType["orderbook:kr"], ["005930"]);
      assert.deepEqual(byType["personal:order"], ["3"]);
      assert.equal(conn.received.size, 0); // 선언은 한 번
    } finally { stream.close(); }
  }));

test("toss: 연결 뒤 추가 구독은 짧게 모아 전체 집합을 다시 선언한다", () =>
  withServer(async (url, connections) => {
    const stream = new TossMarketStream(opts(url));
    try {
      stream.subscribeTrades(["005930"], () => {});
      stream.connect();
      const conn = await connections.take();
      await conn.received.take();
      stream.subscribeTrades(["000660"], () => {});
      stream.subscribeOrderBook(["000660"], () => {});
      const declaration = JSON.parse(await conn.received.take());
      const byType = Object.fromEntries(declaration.slice(1).map((d: { type: string; codes: string[] }) => [d.type, d.codes]));
      assert.deepEqual(byType["trade:kr"], ["000660", "005930"]);
      assert.deepEqual(byType["orderbook:kr"], ["000660"]);
      await sleep(120);
      assert.equal(conn.received.size, 0); // 두 변경이 하나로 합쳐졌다
    } finally { stream.close(); }
  }));

test("toss: 거부된 target 은 다음 선언에서 빠진다", () =>
  withServer(async (url, connections) => {
    const stream = new TossMarketStream(opts(url));
    try {
      stream.subscribeTrades(["005930", "999999"], () => {});
      stream.connect();
      const conn = await connections.take();
      await conn.received.take();
      conn.send('{"type":"subscriptions","id":"req-1","subscribed":["trade:kr:005930"],"rejected":[{"target":"trade:kr:999999","code":"stock-not-found","message":"해당 종목을 찾을 수 없습니다."}]}');
      await sleep(20);
      stream.subscribeTrades(["000660"], () => {});
      const declaration = JSON.parse(await conn.received.take());
      assert.deepEqual(declaration[1], { type: "trade:kr", codes: ["000660", "005930"] });
    } finally { stream.close(); }
  }));

test("toss: 체결·호가 프레임을 요청 표기 심볼로 전달한다 (KRX·US, 접두 없이 구독하면 접두 없이)", () =>
  withServer(async (url, connections) => {
    const ticks = new Queue<TradeTick>();
    const books = new Queue<OrderBookTick>();
    const stream = new TossMarketStream(opts(url));
    try {
      stream.subscribeTrades(["US:AAPL", "005930"], (t) => ticks.put(t));
      stream.subscribeOrderBook(["KRX:005930"], (b) => books.put(b));
      stream.connect();
      const conn = await connections.take();
      await conn.received.take();
      conn.send(TRADE_US);
      const us = await ticks.take();
      assert.equal(us.symbol, "US:AAPL");
      assert.ok(us.price.eq("243.26"));
      assert.ok(us.quantity.eq(8));
      assert.equal(us.cumulativeVolume, null);
      conn.send(TRADE_KR);
      const kr = await ticks.take();
      assert.equal(kr.symbol, "005930"); // 접두 없이 구독 → 접두 없이
      assert.ok(kr.price.eq(71500));
      conn.send(ORDERBOOK_KR);
      const book = await books.take();
      assert.equal(book.symbol, "005930"); // 같은 topic 은 먼저 구독한 표기("005930")로 돌려준다 (Kotlin 과 동일)
      assert.ok(book.asks[0].price.eq(71500));
      assert.ok(book.asks[0].quantity.eq(5));
      assert.ok(book.bids[0].price.eq(71400));
      assert.ok(book.bids[0].quantity.eq(10));
    } finally { stream.close(); }
  }));

test("toss: 주문 이벤트 - FILL 은 체결량·평균가·잔량, 부분 체결은 누적 차이로 계산한다", () =>
  withServer(async (url, connections) => {
    const events = new Queue<OrderEvent>();
    const stream = new TossMarketStream(opts(url));
    try {
      stream.subscribeOrderEvents((e) => events.put(e));
      stream.connect();
      const conn = await connections.take();
      await conn.received.take();
      conn.send(ORDER_FILL);
      const fill = await events.take();
      assert.equal(fill.type, "FILLED");
      assert.equal(fill.symbol, "US:AAPL");
      assert.equal(fill.side, "BUY");
      assert.ok(fill.quantity!.eq(10));
      assert.ok(fill.price!.eq(100));
      assert.ok(fill.remainingQuantity!.eq(0));

      conn.send(orderFrame("PENDING", "0"));
      assert.equal((await events.take()).type, "ACCEPTED");
      conn.send(orderFrame("PARTIAL_FILL", "4"));
      const partial = await events.take();
      assert.ok(partial.quantity!.eq(4));
      assert.ok(partial.remainingQuantity!.eq(6));
      conn.send(orderFrame("FILL", "10", "100.5", "100.5"));
      const rest = await events.take();
      assert.ok(rest.quantity!.eq(6));
      assert.ok(rest.price!.eq("100.5"));
      assert.ok(rest.remainingQuantity!.eq(0));
      conn.send(orderFrame("CANCELING", "10"));
      conn.send(orderFrame("CANCEL_REJECTED", "10"));
      const rejected = await events.take();
      assert.equal(rejected.type, "REJECTED"); // CANCELING 은 삼켜지고 CANCEL_REJECTED 만 전달
      assert.equal(rejected.reason, "CANCEL_REJECTED");
    } finally { stream.close(); }
  }));

test("toss: 하트비트는 텍스트 PING 을 보내고 pong·error 프레임은 삼킨다", () =>
  withServer(async (url, connections) => {
    const stream = new TossMarketStream(opts(url, { heartbeatMs: 100 }));
    try {
      stream.connect();
      const conn = await connections.take();
      assert.equal(await conn.received.take(), "PING");
      conn.send('{"type":"pong"}');
      conn.send('{"type":"error","error":{"code":"rate-limit-exceeded","message":"too fast"}}');
      assert.equal(await conn.received.take(), "PING");
      assert.equal(stream.isConnected, true);
    } finally { stream.close(); }
  }));

test("toss: 서버가 끊으면 재접속해 전체 집합을 다시 선언한다", () =>
  withServer(async (url, connections) => {
    const stream = new TossMarketStream(opts(url));
    try {
      stream.subscribeTrades(["005930"], () => {});
      stream.connect();
      const first = await connections.take();
      await first.received.take();
      first.socket.close(1000, "server-shutdown");
      const second = await connections.take(10_000);
      assert.equal(second.header("authorization"), "Bearer TOKEN");
      const declaration = JSON.parse(await second.received.take());
      assert.deepEqual(declaration[1], { type: "trade:kr", codes: ["005930"] });
    } finally { stream.close(); }
  }));

test("toss topicKey: 시장 접두 규칙", () => {
  assert.equal(tossTopicKey("KRX:005930"), "kr:005930");
  assert.equal(tossTopicKey("005930"), "kr:005930");
  assert.equal(tossTopicKey("US:aapl"), "us:AAPL");
  assert.throws(() => tossTopicKey("JP:7203"), /지원하지 않는 시장/);
  assert.equal(parseTossTrade("orderbook:kr:005930", {}), null);
  assert.equal(parseTossOrderBook("trade:kr:005930", {}), null);
  assert.equal(parseTossOrderEvent({ event: "REPLACING", order: { orderId: "x" } }, null), null);
  assert.equal(parseTossOrderEvent({ event: "FILL", order: {} }, null), null);
  const modified = parseTossOrderEvent({ event: "REPLACED", order: { orderId: "o", quantity: "5", price: "1", currency: "KRW", symbol: "005930", side: "SELL", execution: { filledQuantity: "0" } } }, null)!;
  assert.equal(modified.type, "MODIFIED");
  assert.equal(modified.symbol, "KRX:005930");
  assert.equal(modified.side, "SELL");
  assert.ok(modified.remainingQuantity!.eq(5));
  // 첫 스냅샷의 누적 체결량이 이전보다 작아도 음수가 되지 않는다
  assert.ok(parseTossOrderEvent({ event: "FILL", order: { orderId: "o", quantity: "5", execution: { filledQuantity: "2" } } }, new Decimal(3))!.quantity!.eq(2));
});

test("toss 픽스처: stream 섹션(AsyncAPI 예시)을 같은 모델로 파싱한다", () => {
  assert.equal(fx.measured, false);
  assertTrades(fx.frames.map((f) => { const n = JSON.parse(f); return parseTossTrade(n.topic, n.data)!; }), fx.expected);
  assert.equal(fx.orderBook.measured, false);
  assertBooks(fx.orderBook.frames.map((f) => { const n = JSON.parse(f); return parseTossOrderBook(n.topic, n.data)!; }), fx.orderBook.expected);
  assert.equal(fx.orderEvents.measured, false);
  assertEvents(fx.orderEvents.frames.map((f) => parseTossOrderEvent(JSON.parse(f).data, null)!), fx.orderEvents.expected);
});
